### PR TITLE
Add user-agent parameter to config

### PR DESCRIPTION
**Pull request by yu-kioo**
## overview
add optional user-agent parameter to JDBC and SDK connection

### CODE REVIEW

File: src/main/java/org/embulk/output/databricks/DatabricksAPIClient.java [hunk 1]
```diff
@@ -11,9 +12,18 @@
...
+    SetUserAgent(task);
+
     return new DatabricksAPIClient(createDatabricksConfig(task));
   }
 
+  private static void SetUserAgent(DatabricksOutputPlugin.DatabricksPluginTask task) {
+    String name = task.getUserAgent().get().get("product_name");
+    String version = task.getUserAgent().get().get("product_version");
+
+    UserAgent.withProduct(name, version);
```
Comment (yu-kioo) [inline src/main/java/org/embulk/output/databricks/DatabricksAPIClient.java:24]: [FYI]
[UserAgent.withProduct: defined here](https://github.com/databricks/databricks-sdk-java/blob/22e2561db98389f7d94fb353696c08709925265b/databricks-sdk-java/src/main/java/com/databricks/sdk/core/UserAgent.java#L37) and [set header in client](https://github.com/databricks/databricks-sdk-java/blob/22e2561db98389f7d94fb353696c08709925265b/databricks-sdk-java/src/main/java/com/databricks/sdk/core/ApiClient.java#L256-L257)

File: src/main/java/org/embulk/output/databricks/DatabricksAPIClient.java [hunk 2]
```diff
@@ -11,9 +12,18 @@
 
 public class DatabricksAPIClient {
   public static DatabricksAPIClient create(DatabricksOutputPlugin.DatabricksPluginTask task) {
+    SetUserAgent(task);
+
     return new DatabricksAPIClient(createDatabricksConfig(task));
   }
 
+  private static void SetUserAgent(DatabricksOutputPlugin.DatabricksPluginTask task) {
```
Comment (NamedPython) [inline src/main/java/org/embulk/output/databricks/DatabricksAPIClient.java:20]: 普通にメソッドなので先頭は小文字でいい気がします！僕の知らないお作法だったらごめんなさい！

```suggestion
  private static void setUserAgent(DatabricksOutputPlugin.DatabricksPluginTask task) {
```

File: src/main/java/org/embulk/output/DatabricksOutputPlugin.java [hunk 3]
```diff
@@ -52,6 +53,20 @@ public interface DatabricksPluginTask extends PluginTask {
     @Config("delete_stage_on_error")
     @ConfigDefault("false")
     public boolean getDeleteStageOnError();
+
+    @Config("user_agent")
+    @ConfigDefault("{}")
+    public Optional<UserAgentEntry> getUserAgent();
```
Comment (d-hrs) [inline src/main/java/org/embulk/output/DatabricksOutputPlugin.java:59]:
```suggestion
    public UserAgentEntry getUserAgentEntry();
```

- If there is a default value, then `Optional` is not needed.
- If there is no special reason, it should be the same as the option name.

**Review by d-hrs — CHANGES_REQUESTED** [review]
please check my comment.

File: src/main/java/org/embulk/output/DatabricksOutputPlugin.java (hunk 3 above)
Comment (yu-kioo) [reply src/main/java/org/embulk/output/DatabricksOutputPlugin.java]: fixed in [this commit](https://github.com/trocco-io/embulk-output-databricks/pull/4/commits/0e4ed29571b0345865c38361ed630f50a3e3d09f).

File: src/main/java/org/embulk/output/DatabricksOutputPlugin.java [hunk 4]
```diff
@@ -92,6 +107,14 @@ protected JdbcOutputConnector getConnector(PluginTask task, boolean retryableMet
     props.put("ConnCatalog", t.getCatalogName());
     props.put("ConnSchema", t.getSchemaName());
     props.putAll(t.getOptions());
+    // overwrite UserAgentEntry property if the same property is set in options
+    if (t.getUserAgentEntry() != null) {
```
Comment (chikamura) [inline src/main/java/org/embulk/output/DatabricksOutputPlugin.java:111]: `t.getUserAgentEntry` always seems to be not null, May I suggest that you can delete this line.

**Review by chikamura — CHANGES_REQUESTED** [review]

File: src/main/java/org/embulk/output/DatabricksOutputPlugin.java [hunk 5]
```diff
@@ -92,6 +107,14 @@ protected JdbcOutputConnector getConnector(PluginTask task, boolean retryableMet
...
+      String product_name = t.getUserAgentEntry().getProductName();
+      String product_version = t.getUserAgentEntry().getProductVersion();
+
+      props.put("UserAgentEntry", product_name + "/" + product_version);
```
Comment (chikamura) [inline src/main/java/org/embulk/output/DatabricksOutputPlugin.java:115]: Java's variable names are usually camel case.

product_name -> productName
product_version -> productVersion

File: src/main/java/org/embulk/output/DatabricksOutputPlugin.java (hunk 4 above)
Comment (yu-kioo) [reply src/main/java/org/embulk/output/DatabricksOutputPlugin.java]: The current code did not guarantee the behavior of 2 below, so changed condition.

- 1: options is not set & user-agent is not set => unknown/0.0.0
- 2: options is set & user-agent is not set => value of options
- 3: options is not set & user-agent is set => value of user-agent
- 4: options is set, user-agent is set => value of user-agent

File: src/main/java/org/embulk/output/DatabricksOutputPlugin.java (hunk 5 above)
Comment (yu-kioo) [reply src/main/java/org/embulk/output/DatabricksOutputPlugin.java]: sorry, fixed.

File: src/main/java/org/embulk/output/DatabricksOutputPlugin.java (hunk 4 above)
Comment (yu-kioo) [reply src/main/java/org/embulk/output/DatabricksOutputPlugin.java]: We have decided to always overwrite the UserAgentEntry value in options by user-agent, so please ignore the above comment. I removed condition as you commented.

**Review by chikamura — APPROVED** [review]
LGTM

**Pull request merged**